### PR TITLE
DAOS-7371 test: Objects are scrubbed properly by scrubber.

### PR DESCRIPTION
**Pull request by rpadma2**
Test-tag: test_objects_scrubbed_properly always_passes test_tiers

Summary:
   - Update the scrub-freq and validate objects are scrubbed properly.
   - Set the initial scrub-freq to default value (ie: 1 week)
   - Reset the scrub-freq to 5 seconds during testing.

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Comment by rpadma2** [comment]
All new tests passed.

File: src/tests/ftest/scrubber/frequency.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+            1. Create a test pool without scrubber properties.
+               The default scrub frequency is set to 1 week.
+            2. Initial scrub happens when the objects are created.
+            3. Check objects are not scrubbed after initial scan.
+               Wait for 5 minutes and see scrubber not doing
+               any more object scrubbing now.
+            4. Update the scrubber frequency to 5 seconds.
+            5. Observe the objects are scrubbed every 5 seconds.
+
+        :avocado: tags=all,pr,daily_regression
```
Comment (daltonbohning) [inline src/tests/ftest/scrubber/frequency.py:30]: pr and daily?

File: src/tests/ftest/scrubber/frequency.yaml [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        - DD_MASK=mgmt,md,rebuild
+      storage: auto
+  transport_config:
+    allow_insecure: true
+agent_config:
+  transport_config:
+    allow_insecure: true
+dmg:
+  transport_config:
+    allow_insecure: true
```
Comment (daltonbohning) [inline src/tests/ftest/scrubber/frequency.yaml:38]:
```suggestion
```

File: src/tests/ftest/scrubber/frequency.yaml [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+pool:
+  scm_size: 6000000000
+  nvme_size: 54000000000
+  svcn: 4
+  control_method: dmg
```
Comment (daltonbohning) [inline src/tests/ftest/scrubber/frequency.yaml:43]:
```suggestion
```

File: src/tests/ftest/scrubber/frequency.yaml [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  properties: "scrub:timed"
+container:
+  type: POSIX
+  control_method: daos
+  oclass: RP_2G1
+ior:
+  ior_timeout: 60
+  client_processes:
+    np_16:
+      np: 2
```
Comment (daltonbohning) [inline src/tests/ftest/scrubber/frequency.yaml:55]:
```suggestion
  client_processes:
    np: 2
```

File: src/tests/ftest/scrubber/frequency.yaml [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+  dfs_destroy: false
+  iorflags:
+    ior_flags:
+      - "-v -W -w -r -R -k"
+    ior_api:
+      - DFS
+    transfer_block_size:
+      - [1M, 32M]
+    obj_class:
+      - RP_2G1
```
Comment (daltonbohning) [inline src/tests/ftest/scrubber/frequency.yaml:67]: We should simplify this
```suggestion
  flags: "-v -W -w -r -R -k"
  api: DFS
  transfer_size: 1M
  block_size: 32M
  dfs_oclass: RP_2G1
```

File: src/tests/ftest/scrubber/frequency.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        :avocado: tags=hw,medium
+        :avocado: tags=scrubber
+        :avocado: tags=TestWithScrubberFreq,test_objects_scrubbed_properly
+        """
+        flags = self.params.get("ior_flags", '/run/ior/iorflags/*')
+        apis = self.params.get("ior_api", '/run/ior/iorflags/*')
+        transfer_block_size = self.params.get("transfer_block_size",
+                                              '/run/ior/iorflags/*')
+        obj_class = self.params.get("obj_class", '/run/ior/iorflags/*')
```
Comment (daltonbohning) [inline src/tests/ftest/scrubber/frequency.py:39]: With recommended config changes, this isn't needed
```suggestion
```

File: src/tests/ftest/scrubber/frequency.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+        pool_prop = self.params.get("properties", '/run/pool/*')
+        cont_prop = self.params.get("properties", '/run/container/*')
+        self.ior_cmd.api.update(apis[0])
+        self.ior_cmd.flags.update(flags[0], "ior.flags")
+        self.ior_cmd.dfs_oclass.update(obj_class[0])
+        self.ior_cmd.dfs_dir_oclass.update(obj_class[0])
```
Comment (daltonbohning) [inline src/tests/ftest/scrubber/frequency.py:45]: ... or this
```suggestion
```

File: src/tests/ftest/scrubber/frequency.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+        self.create_pool_cont_with_scrubber(pool_prop=pool_prop, cont_prop=cont_prop)
+        for test in transfer_block_size:
+            self.ior_cmd.transfer_size.update(test[0])
+            self.ior_cmd.block_size.update(test[1])
```
Comment (daltonbohning) [inline src/tests/ftest/scrubber/frequency.py:49]: ... or this
```suggestion
```

File: src/tests/ftest/scrubber/frequency.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+        self.log.info("Final scrubber metrics with scrub:timed")
+        final_scrubbed_metrics = self.scrubber.get_scrubber_bytes_scrubbed_total()
+        status = self.verify_scrubber_metrics_value(initial_scrubbed_metrics,
+                                                    final_scrubbed_metrics)
+        if status is True:
+            self.fail("--Test Failed: Metrics Value is Changing--")
+        # Now set the scrub-freq to 5 seconds.
+        self.pool.set_property("scrub-freq", "5")
+        self.log.info("Sleeping for 5 secs pool property scrub:timed,scrub-freq:5")
+        sleep(5)
```
Comment (shimizukko) [inline src/tests/ftest/scrubber/frequency.py:69]: Why do we sleep 5 sec here? If the scrubbing happens during this 5 sec, wouldn't `initial_scrubbed_metrics` and `final_scrubbed_metrics` become equal? What if we obtain `initial_scrubbed_metrics` before setting the frequency to 5 sec?

**Review by shimizukko — APPROVED** [review]

File: src/tests/ftest/scrubber/frequency.py (hunk 7 above)
Comment (rpadma2) [reply src/tests/ftest/scrubber/frequency.py:45]: Will do.

File: src/tests/ftest/scrubber/frequency.yaml (hunk 4 above)
Comment (rpadma2) [reply src/tests/ftest/scrubber/frequency.yaml:55]: Will do.

File: src/tests/ftest/scrubber/frequency.yaml (hunk 3 above)
Comment (rpadma2) [reply src/tests/ftest/scrubber/frequency.yaml:43]: Yes. This is the problem when you start the initial template with cut and paste :-)

File: src/tests/ftest/scrubber/frequency.py (hunk 8 above)
Comment (daltonbohning) [reply src/tests/ftest/scrubber/frequency.py]: Thanks :)

File: src/tests/ftest/scrubber/frequency.py [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+        :avocado: tags=all,daily_regression
+        :avocado: tags=hw,medium
+        :avocado: tags=scrubber
+        :avocado: tags=TestWithScrubberFreq,test_objects_scrubbed_properly
+        """
+        pool_prop = self.params.get("properties", '/run/pool/*')
+        cont_prop = self.params.get("properties", '/run/container/*')
+        self.create_pool_cont_with_scrubber(pool_prop=pool_prop, cont_prop=cont_prop)
+        # Run IOR and gather the total scrubbed metrics information.
+        self.run_ior_and_check_scruber_status(pool=self.pool, cont=self.container)
```
Comment (daltonbohning) [inline src/tests/ftest/scrubber/frequency.py:39]: FYI this pre-existing function has a typo
```suggestion
        self.run_ior_and_check_scrubber_status(pool=self.pool, cont=self.container)
```

**Review by daltonbohning — DISMISSED** [review]

File: src/tests/ftest/scrubber/frequency.py (hunk 10 above)
Comment (rpadma2) [reply src/tests/ftest/scrubber/frequency.py:39]: Great... You spotted it... Sigh... Will push the changes.

Comment (daltonbohning) [reply src/tests/ftest/scrubber/frequency.py:39]: lol

Comment (daltonbohning) [reply src/tests/ftest/scrubber/frequency.py:39]: I think it was the mdtest one that you fixed before.

Comment (rpadma2) [reply src/tests/ftest/scrubber/frequency.py:39]: The problem was I changed the implementation on the final mdtest script (Which was checked in) without interfering with the scrubber_test_base.py file (where the typo error was there).

**Review by shimizukko — APPROVED** [review]

**Comment by rpadma2** [comment]
@daos-stack/daos-gatekeeper  ; This PR is ready for merge. All necessary tests passed.

**Pull request merged**